File: pkg/wire/activity.go
package wire

// Activity journal shapes (docs/COMMANDER.md §3). One set of structs is
// shared by the router's store, the SDK's Note, and the shell verbs, so
// an entry that an app noted is byte-for-byte the entry a Timeline row
// renders.

// ActivityIntent is the way back to what an entry describes — exactly the
// shapes the start menu's Recent pop-outs and the agent verbs already
// accept, so a consumer needs no new plumbing to act on a row.
type ActivityIntent struct {
	// Kind is focus | resume | open.
	Kind string `json:"kind"`
	// Origin names the host for focus; empty means the entry's own.
	Origin     string `json:"origin,omitempty"`
	AppID      string `json:"app_id,omitempty"`
	InstanceID string `json:"instance_id,omitempty"`
	WindowID   uint32 `json:"window_id,omitempty"`
	SessionID  string `json:"session_id,omitempty"`
	RowKey     string `json:"row_key,omitempty"`
	Path       string `json:"path,omitempty"`
}

// ActivityEntry is one journal row. (Host, Seq) is its identity.
type ActivityEntry struct {
	TS   int64  `json:"ts"`
	Seq  uint64 `json:"seq"`
	Host string `json:"host"`
	// Kind is dotted and namespaced by source: window.open, open.routed,
	// session.attach, agent.turn, priv.escalate, brief, rollup …
	Kind     string `json:"kind"`
	App      string `json:"app,omitempty"`
	Instance string `json:"instance,omitempty"`
	Window   uint32 `json:"window,omitempty"`
	Title    string `json:"title,omitempty"`
	// Line is plain text, bounded by the store; Truncated says it was cut.
	// Ref is source-defined (a transcript seq, a path) and small.
	Line      string          `json:"line"`
	Truncated bool            `json:"truncated,omitempty"`
	Ref       map[string]any  `json:"ref,omitempty"`
	Intent    *ActivityIntent `json:"intent,omitempty"`
}

// ActivityStats is what About shows: what is stored, and what was dropped.
type ActivityStats struct {
	Enabled bool           `json:"enabled"`
	Days    int            `json:"days"`
	Bytes   int64          `json:"bytes"`
	Dropped uint64         `json:"dropped"`
	Today   map[string]int `json:"today"`
	Seq     uint64         `json:"seq"`
	Path    string         `json:"path,omitempty"`
}

// --- app → router ---

// TEvtActivityNote: an app backend records a fact about itself. The router
// stamps time, host, app id and instance id from the attested sender; a
// note that names a window the instance does not own, or arrives faster
// than the per-instance limit, is dropped and logged. Requires
// CapActivityNote. Fire-and-forget.
const TEvtActivityNote = "activity.note"

// EvtActivityNote is the app half of an ActivityEntry: everything the
// router cannot know itself.
type EvtActivityNote struct {
	T      string          `json:"t"`
	Kind   string          `json:"kind"`
	Win    uint32          `json:"win,omitempty"`
	Title  string          `json:"title,omitempty"`
	Line   string          `json:"line"`
	Ref    map[string]any  `json:"ref,omitempty"`
	Intent *ActivityIntent `json:"intent,omitempty"`
}

func NewEvtActivityNote(kind, line string) EvtActivityNote {
	return EvtActivityNote{T: TEvtActivityNote, Kind: kind, Line: line}
}

// --- shell ↔ router ---

const (
	// Shell → router: a page of the journal, newest first.
	TShellActivityQuery = "activity.query"
	// Router → shell: the page. Cursor is set when more entries precede it.
	TShellActivityQueryOK = "activity.query.ok"
	// Router → shell: the query was refused (journal off, bad cursor).
	TShellActivityQueryErr = "activity.query.err"
	// Shell → router: start (On) or stop pushing new entries as they land.
	TShellActivityTail = "activity.tail"
	// Router → shell: one new entry, telemetry class, while a tail is on.
	TShellActivityEntry = "activity.entry"
	// Shell → router: the store's footprint. Router → shell: the answer.
	TShellActivityStats   = "activity.stats"
	TShellActivityStatsOK = "activity.stats.ok"
	// Shell → router: delete every day file. Router → shell: done.
	TShellActivityClear   = "activity.clear"
	TShellActivityClearOK = "activity.clear.ok"
)

type ShellActivityQuery struct {
	T     string `json:"t"`
	ReqID uint64 `json:"req_id"`
	// From/To bound the entry time (unix ms), inclusive; zero is open.
	From int64 `json:"from,omitempty"`
	To   int64 `json:"to,omitempty"`
	// Kinds match exactly, or by prefix when they end in "."; Apps match
	// exactly; Text is a case-insensitive substring over title and line.
	Kinds  []string `json:"kinds,omitempty"`
	Apps   []string `json:"apps,omitempty"`
	Text   string   `json:"text,omitempty"`
	Limit  int      `json:"limit,omitempty"`
	Cursor string   `json:"cursor,omitempty"`
}

func NewShellActivityQuery(reqID uint64) ShellActivityQuery {
	return ShellActivityQuery{T: TShellActivityQuery, ReqID: reqID}
}

type ShellActivityQueryOK struct {
	T       string          `json:"t"`
	ReqID   uint64          `json:"req_id"`
	Entries []ActivityEntry `json:"entries"`
	Cursor  string          `json:"cursor,omitempty"`
}

func NewShellActivityQueryOK(reqID uint64, entries []ActivityEntry, cursor string) ShellActivityQueryOK {
	if entries == nil {
		entries = []ActivityEntry{}
	}
	return ShellActivityQueryOK{T: TShellActivityQueryOK, ReqID: reqID, Entries: entries, Cursor: cursor}
}

type ShellActivityQueryErr struct {
	T     string `json:"t"`
	ReqID uint64 `json:"req_id"`
	Code  string `json:"code"`
	Msg   string `json:"msg,omitempty"`
}

func NewShellActivityQueryErr(reqID uint64, code, msg string) ShellActivityQueryErr {
	return ShellActivityQueryErr{T: TShellActivityQueryErr, ReqID: reqID, Code: code, Msg: msg}
}

type ShellActivityTail struct {
	T  string `json:"t"`
	On bool   `json:"on"`
}

func NewShellActivityTail(on bool) ShellActivityTail {
	return ShellActivityTail{T: TShellActivityTail, On: on}
}

type ShellActivityEntry struct {
	T     string        `json:"t"`
	Entry ActivityEntry `json:"entry"`
}

func NewShellActivityEntry(e ActivityEntry) ShellActivityEntry {
	return ShellActivityEntry{T: TShellActivityEntry, Entry: e}
}

type ShellActivityStats struct {
	T     string `json:"t"`
	ReqID uint64 `json:"req_id"`
}

func NewShellActivityStats(reqID uint64) ShellActivityStats {
	return ShellActivityStats{T: TShellActivityStats, ReqID: reqID}
}

type ShellActivityStatsOK struct {
	T     string        `json:"t"`
	ReqID uint64        `json:"req_id"`
	Stats ActivityStats `json:"stats"`
}

func NewShellActivityStatsOK(reqID uint64, st ActivityStats) ShellActivityStatsOK {
	return ShellActivityStatsOK{T: TShellActivityStatsOK, ReqID: reqID, Stats: st}
}

type ShellActivityClear struct {
	T     string `json:"t"`
	ReqID uint64 `json:"req_id"`
}

func NewShellActivityClear(reqID uint64) ShellActivityClear {
	return ShellActivityClear{T: TShellActivityClear, ReqID: reqID}
}

type ShellActivityClearOK struct {
	T     string `json:"t"`
	ReqID uint64 `json:"req_id"`
}

func NewShellActivityClearOK(reqID uint64) ShellActivityClearOK {
	return ShellActivityClearOK{T: TShellActivityClearOK, ReqID: reqID}
}

File: web/shell/src/activity.ts
// activity — the shell's side of the router's activity journal
// (docs/COMMANDER.md §3). Query, stats and clear are req_id-correlated
// promises over the ctrl channel, one per origin; the tail is a push the
// router sends while it is on, fanned out here to subscribers.
//
// Every host journals itself, so an entry's `host` is stamped HERE from the
// origin it arrived on: a remote router calls itself "local" too.
//
// Wire (pkg/wire/activity.go):
//   shell → router: { t: "activity.query", req_id, from?, to?, kinds?, apps?, text?, limit?, cursor? }
//   router → shell: { t: "activity.query.ok", req_id, entries, cursor? } | { t: "activity.query.err", req_id, code, msg }
//   shell → router: { t: "activity.tail", on }      router → shell: { t: "activity.entry", entry }
//   shell → router: { t: "activity.stats", req_id } router → shell: { t: "activity.stats.ok", req_id, stats }
//   shell → router: { t: "activity.clear", req_id } router → shell: { t: "activity.clear.ok", req_id }

import { LOCAL_ORIGIN, type Origin } from './clients';

export interface ActivityIntent {
  kind: 'focus' | 'resume' | 'open' | string;
  origin?: string;
  app_id?: string;
  instance_id?: string;
  window_id?: number;
  session_id?: string;
  row_key?: string;
  path?: string;
}

export interface ActivityEntry {
  ts: number;
  seq: number;
  /** the origin this entry came from ('local' for the seat's own host) */
  host: string;
  kind: string;
  app?: string;
  instance?: string;
  window?: number;
  title?: string;
  line: string;
  truncated?: boolean;
  ref?: Record<string, unknown>;
  intent?: ActivityIntent;
}

export interface ActivityStats {
  enabled: boolean;
  days: number;
  bytes: number;
  dropped: number;
  today: Record<string, number>;
  seq: number;
  path?: string;
}

export interface ActivityQuery {
  from?: number;
  to?: number;
  kinds?: string[];
  apps?: string[];
  text?: string;
  limit?: number;
  cursor?: string;
}

export interface ActivityPage {
  host: string;
  entries: ActivityEntry[];
  cursor?: string;
}

type SendCtrl = (msg: unknown) => void;

interface Pending {
  resolve: (v: unknown) => void;
  reject: (err: Error) => void;
  origin: Origin;
}

const pending = new Map<number, Pending>();
let nextReqID = 1;

function hostOf(origin: Origin): string {
  return origin === LOCAL_ORIGIN ? 'local' : origin;
}

function stamp(origin: Origin, e: ActivityEntry): ActivityEntry {
  return { ...e, host: hostOf(origin) };
}

function request<T>(send: SendCtrl, origin: Origin, msg: Record<string, unknown>): Promise<T> {
  const reqID = nextReqID++;
  return new Promise<T>((resolve, reject) => {
    pending.set(reqID, { resolve: resolve as (v: unknown) => void, reject, origin });
    send({ ...msg, req_id: reqID });
  });
}

/** activityQuery asks one origin's router for a page of its journal. */
export function activityQuery(send: SendCtrl, origin: Origin, q: ActivityQuery): Promise<ActivityPage> {
  return request<ActivityPage>(send, origin, { t: 'activity.query', ...q });
}

export function activityStats(send: SendCtrl, origin: Origin): Promise<ActivityStats> {
  return request<ActivityStats>(send, origin, { t: 'activity.stats' });
}

export function activityClear(send: SendCtrl, origin: Origin): Promise<void> {
  return request<void>(send, origin, { t: 'activity.clear' });
}

// ---- replies, called from main.tsx's ctrl dispatcher ----

export function handleActivityQueryOK(origin: Origin, msg: { req_id: number; entries: ActivityEntry[]; cursor?: string }): void {
  const p = pending.get(msg.req_id);
  if (!p) return;
  pending.delete(msg.req_id);
  p.resolve({ host: hostOf(origin), entries: (msg.entries ?? []).map((e) => stamp(origin, e)), cursor: msg.cursor } satisfies ActivityPage);
}

export function handleActivityQueryErr(msg: { req_id: number; code: string; msg?: string }): void {
  const p = pending.get(msg.req_id);
  if (!p) return;
  pending.delete(msg.req_id);
  p.reject(new Error(`${msg.code}${msg.msg ? ': ' + msg.msg : ''}`));
}

export function handleActivityStatsOK(origin: Origin, msg: { req_id: number; stats: ActivityStats }): void {
  const p = pending.get(msg.req_id);
  if (!p) return;
  pending.delete(msg.req_id);
  p.resolve({ ...msg.stats, path: msg.stats.path });
  void origin;
}

export function handleActivityClearOK(msg: { req_id: number }): void {
  const p = pending.get(msg.req_id);
  if (!p) return;
  pending.delete(msg.req_id);
  p.resolve(undefined);
}

/** rejectPendingFor fails every request waiting on an origin whose
 *  connection went away, so a Timeline does not wait for ever. */
export function rejectPendingFor(origin: Origin): void {
  for (const [id, p] of pending) {
    if (p.origin === origin) {
      pending.delete(id);
      p.reject(new Error('disconnected'));
    }
  }
}

// ---- tail ----

const subscribers = new Set<(e: ActivityEntry) => void>();

/** onActivity subscribes to live entries from every origin. The caller
 *  (main.tsx) turns the routers' tails on while anyone is subscribed. */
export function onActivity(cb: (e: ActivityEntry) => void): () => void {
  subscribers.add(cb);
  return () => { subscribers.delete(cb); };
}

export function tailWanted(): boolean {
  return subscribers.size > 0;
}

export function handleActivityEntry(origin: Origin, msg: { entry: ActivityEntry }): void {
  const e = stamp(origin, msg.entry);
  for (const cb of subscribers) cb(e);
}
